test/unit/helpers/foundation.js

```diff
@@ -24,44 +24,132 @@
+ * preferred way of testing interaction handlers.
+ *
+ * ```javascript
+ * test('#init adds a click listener which adds a "foo" class', (t) => {
+ *   const {foundation, mockAdapter} = setupTest();
+ *   const handlers = captureHandlers(mockAdapter, 'registerInteractionHandler');
+ *   foundation.init();
+ *   handlers.click(/* you can pass event info in here *\/ {type: 'click'});
+ *   t.doesNotThrow(() => td.verify(mockAdapter.addClass('foo')));
+ *   t.end();
+ * });
+ * ```
+ *
+ * Note that `handlerCaptureMethodName` _must_ have a signature of `(string, EventListener) => any` in order to
+ * be effective.
+ *
+ * @param {!A} adapter
+ * @param {string} handlerCaptureMethodName
+ * @template A
+ */
+export function captureHandlers(adapter, handlerCaptureMethodName) {
   const {isA} = td.matchers;
   const handlers = {};
-  td.when(adapter[handlerCaptureMethod](isA(String), isA(Function))).thenDo((type, handler) => {
+  td.when(adapter[handlerCaptureMethodName](isA(String), isA(Function))).thenDo((type, handler) => {
     handlers[type] = (evtInfo = {}) => handler(Object.assign({type}, evtInfo));
   });
   return handlers;
 }
+
+/**
+ * @param {!Array<string>} actualArray
+ * @param {!Array<string>} expectedArray
+ * @return {string}
+ */
+function getUnequalArrayMessage(actualArray, expectedArray) {
+  /**
+   * @param {!Array<string>} values
+   * @param {string} singularName
+   * @return {string}
+   */
+  const format = (values, singularName) => {
+    const count = values.length;
+    if (count === 0) {
+      return '';
+    }
+    const plural = count === 1 ? '' : 's';
+    const str = values.join(', ');
+    return `${count} ${singularName}${plural}: ${str}`;
+  };
+
+  /**
+   * @param {!Set<string>} actualSet
+   * @param {!Set<string>} expectedSet
+   */
+  const getAddedStr = (actualSet, expectedSet) => {
+    const addedArray = [];
+    actualSet.forEach((val) => {
+      if (!expectedSet.has(val)) {
+        addedArray.push(val);
+      }
+    });
+    return format(addedArray, 'unexpected method');
+  };
+
+  /**
+   * @param {!Set<string>} actualSet
+   * @param {!Set<string>} expectedSet
+   */
+  const getRemovedStr = (actualSet, expectedSet) => {
+    const removedArray = [];
+    expectedSet.forEach((val) => {
+      if (!actualSet.has(val)) {
+        removedArray.push(val);
+      }
+    });
+    return format(removedArray, 'missing method');
+  };
+
+  /**
+   * @param {!Array<string>} array
+   * @return {!Set<string>}
+   */
+  const toSet = (array) => {
+    const set = new Set();
+    array.forEach((value) => set.add(value));
+    return set;
+  };
+
+  const actualSet = toSet(actualArray);
+  const expectedSet = toSet(expectedArray);
+  const addedStr = getAddedStr(actualSet, expectedSet);
+  const removedStr = getRemovedStr(actualSet, expectedSet);
+  const messages = [addedStr, removedStr].filter((val) => Boolean(val));
```

Comment:
`(val) => val.length > 0` ? Makes it a little more clear this is just filtering out empty strings...

Reply:
Done